### PR TITLE
Allow to specify custom attributes to update on conflict for upsert_all method

### PR DESCRIPTION
**Pull request by simi**
This allows to specify attributes to update on conflict. It opens door to handling timestamps updates manually in reasonable transparent way.

```ruby
now = Time.now
row = { slug: 'new-title', post: 'New post', created_at: now, updated_at: now }
Post.upsert(row1, unique_by: { columns: [:slug]}, update: [:post, :updated_at])
```

In combination with defaults (I'll try to add this in different PR) mentioned at https://github.com/rails/rails/issues/35493#issuecomment-471526290 it can provide easy and transparent way to handle timestamps. Anyway I think it should not be default behaviour since methods like `update_all` doesn't handle this on its own as well.

WDYT @boblail @dhh @kaspth :question:

references https://github.com/rails/rails/issues/35493 https://github.com/rails/rails/pull/35494

I'll update documentation once approved.

### CODE REVIEW

**Comment by simi** [comment]
@boblail ping

File: activerecord/lib/active_record/insert_all.rb [hunk 1]
```diff
@@ -25,7 +27,7 @@ def keys
       inserts.first.keys.map(&:to_s)
     end
 
-    def updatable_columns
+    def default_updatable_columns
```
Comment (boblail) [inline activerecord/lib/active_record/insert_all.rb]: In the context of this PR, I like the rename of `updatable_columns` to `default_updatable_columns`.

File: activerecord/lib/active_record/persistence.rb [hunk 2]
```diff
@@ -254,8 +254,8 @@ def upsert(attributes, returning: nil, unique_by: nil)
       #     { title: 'Clean Code', author: 'Robert', isbn: '2' }
       #   ], unique_by: { columns: %w[ isbn ] })
       #
-      def upsert_all(attributes, returning: nil, unique_by: nil)
-        InsertAll.new(self, attributes, on_duplicate: :update, returning: returning, unique_by: unique_by).execute
+      def upsert_all(attributes, returning: nil, unique_by: nil, updatable_columns: nil)
+        InsertAll.new(self, attributes, on_duplicate: :update, returning: returning, unique_by: unique_by, updatable_columns: updatable_columns).execute
```
Comment (boblail) [inline activerecord/lib/active_record/persistence.rb]: For the name of the new keywords, though, `updatable_columns:` seems a bit verbose.

I'd suggest, simply, `:update`.

More conversation here https://github.com/rails/rails/pull/35077#issuecomment-471826877 and starting here https://github.com/rails/rails/pull/35636#issuecomment-473481908

**Comment by simi** [comment]
@boblail :eyes: requested changes addressed.

**Comment by boblail** [comment]
LGTM!

I'm curious what Rails Members think of the feature and of unifying with #35636!

**Comment by simi** [comment]
I think `update` option can work with array of columns or with SQL string (similar to `joins` for example).

**Comment by kaspth** [comment]
I'm not liking the extra API exposed here, doesn't match Rails' aesthetic. Let's continue in #35636

**Comment by simi** [comment]
@kaspth is it only about the API or doesn't you like option to specify which columns to use on update without need of using SQL fragment? In my eyes that is actually really handy and powerful. I can wait until #35636 is resolved and I'll try to rethink this agian. Anyway thanks for your time you spend on this already.

**Comment by kaspth** [comment]
I want to do something closer to only/with that I alluded to in #35686.

**Comment by simi** [comment]
It is hard to follow for me @kaspth since #35686 was closed.

**Comment by kaspth** [comment]
I meant this part from the description:

```ruby
Book.upsert_all rows, only: %i( id name author_id )
Book.upsert_all rows, with: "status = GREATEST(books.status, excluded.status)"
```

Though I'm thinking of putting a halt on more changes for `insert_all` for Rails 6. I want the relation support, but that's probably it for now for me.

**Comment by simi** [comment]
:+1:  Ahh, ok. I'll focus on relation support for now. We can reopen and rethink specifying custom attrobutes to update on conflict once 6 is released and get more time unless next release.

**Comment by shaicoleman** [comment]
@simi , @kaspth , can you look into reopening this?
